### PR TITLE
Limit size of presence EDUs

### PR DESCRIPTION
**Pull request by erikjohnston**
Otherwise they are unbounded.

### CODE REVIEW

**Comment by richvdh** [comment]
Should we have something in the spec about this, so that federating servers can safely apply a request size limit without breaking things?

**Comment by erikjohnston** [comment]
> Should we have something in the spec about this, so that federating servers can safely apply a request size limit without breaking things?

Yes, we should mandate a limit on the request size

File: synapse/federation/sender/per_destination_queue.py [hunk 1]
```diff
@@ -721,22 +722,23 @@ async def __aenter__(self) -> Tuple[List[EventBase], List[Edu]]:
 
         # Add presence EDU.
         if self.queue._pending_presence:
+            # Only send max 50 presence entries in the EDU, to bound the amount
+            # of data we're sending.
+            presence_to_add: List[JsonDict] = []
+            while self.queue._pending_presence and len(presence_to_add) < 50:
```
Comment (anoadragon453) [inline synapse/federation/sender/per_destination_queue.py:728]: Could you extract this into a global somewhere?

File: changelog.d/17371.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Limit size of presence EDUs.
```
Comment (anoadragon453) [inline changelog.d/17371.misc:1]:
```suggestion
Limit size of presence EDUs to 50 entries.
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
LGTM thanks. Could you add a quick unit test to test the limiting behaviour?

**Review by anoadragon453 — APPROVED** [review]
Looks great, thank you!

**Comment by ansiwen** [comment]
This didn't fix the issue for us and others, see https://github.com/element-hq/synapse/issues/17035#issuecomment-2307156054 and https://github.com/element-hq/synapse/issues/17035#issuecomment-2351029837. I could mitigate the issue by going down with the limit to 25, but I think we need a real size limit, either per event or at least per EDU.

**Pull request merged**